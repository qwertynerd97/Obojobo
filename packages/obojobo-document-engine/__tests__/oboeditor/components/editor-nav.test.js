import { mount } from 'enzyme'
import React from 'react'
import renderer from 'react-test-renderer'

import EditorNav from 'src/scripts/oboeditor/components/editor-nav'

import EditorUtil from 'src/scripts/oboeditor/util/editor-util'
jest.mock('src/scripts/oboeditor/util/editor-util')
import ClipboardUtil from 'src/scripts/oboeditor/util/clipboard-util'
jest.mock('src/scripts/oboeditor/util/clipboard-util')
// SubMenu
jest.mock('src/scripts/oboeditor/components/sub-menu')

describe('EditorNav', () => {
	beforeAll(() => {
		document.getSelection = jest.fn()
		document.execCommand = jest.fn()
	})
	beforeEach(() => {
		jest.clearAllMocks()
	})

	test('EditorNav component', () => {
		EditorUtil.getOrderedList.mockReturnValueOnce([])
		const props = {
			navState: {}
		}
		const component = renderer.create(<EditorNav {...props} />)
		const tree = component.toJSON()
		expect(tree).toMatchSnapshot()
	})

	test('EditorNav component with all list item types', () => {
		EditorUtil.getOrderedList.mockReturnValueOnce([
			{
				id: 4,
				type: 'heading',
				label: 'label4'
			},
			{
				id: 56,
				type: 'link',
				label: 'label5',
				contentType: 'Page',
				flags: {
					assessment: false
				}
			},
			{
				id: 6,
				type: 'link',
				label: 'label6',
				flags: {
					assessment: true
				}
			},
			{
				id: 7,
				type: 'sublink',
				label: 'label7',
				flags: {
					assessment: false
				}
			}
		])

		const props = {
			navState: {
				open: false,
				locked: true,
				navTargetId: 56 // select this item
			}
		}
		const component = renderer.create(<EditorNav {...props} />)
		const tree = component.toJSON()
		expect(tree).toMatchSnapshot()
	})

	test('EditorNav component clicks on page item', () => {
		EditorUtil.getOrderedList
			.mockReturnValueOnce([
				{
					id: 6,
					type: 'link',
					label: 'label6',
					flags: {
						assessment: true
					}
				}
			])
			.mockReturnValueOnce([])

		const props = { navState: {} }
		const component = mount(<EditorNav {...props} />)

		component.instance().onClick({
			id: 6,
			type: 'link',
			label: 'label6',
			flags: {
				assessment: true
			}
		})

		expect(EditorUtil.gotoPath).toHaveBeenCalled()
	})

	test('EditorNav component clicks Add Page button', () => {
		EditorUtil.getOrderedList.mockReturnValueOnce([]).mockReturnValueOnce([])
		jest.spyOn(window, 'prompt')
		window.prompt.mockReturnValueOnce(null)

		const props = { navState: {} }
		const component = mount(<EditorNav {...props} />)

		component
			.find('button')
			.at(0)
			.simulate('click')

		expect(EditorUtil.addPage).toHaveBeenCalled()
	})

	test('EditorNav component clicks Add Assessment button', () => {
		EditorUtil.getOrderedList.mockReturnValueOnce([]).mockReturnValueOnce([])
		jest.spyOn(window, 'prompt')
		window.prompt.mockReturnValueOnce(null)

		const props = { navState: {} }
		const component = mount(<EditorNav {...props} />)

		component
			.find('button')
			.at(1)
			.simulate('click')

		expect(EditorUtil.addAssessment).toHaveBeenCalled()
	})

	test('EditorNav component clicks Rename Module button and cancels change', () => {
		EditorUtil.getOrderedList.mockReturnValueOnce([
			{
				id: 6,
				type: 'heading',
				label: 'label6',
				flags: {
					assessment: true
				}
			}
		])
		jest.spyOn(window, 'prompt')
		window.prompt.mockReturnValueOnce(null)

		const props = { navState: {} }
		const component = mount(<EditorNav {...props} />)

		component
			.find('button')
			.at(2)
			.simulate('click')

		expect(EditorUtil.renamePage).not.toHaveBeenCalled()
	})

	test('EditorNav component clicks Rename Module button with empty name', () => {
		EditorUtil.getOrderedList.mockReturnValueOnce([
			{
				id: 6,
				type: 'heading',
				label: 'label6',
				flags: {
					assessment: true
				}
			}
		])
		jest.spyOn(window, 'prompt')
		window.prompt.mockReturnValueOnce('   ')

		const props = { navState: {} }
		const component = mount(<EditorNav {...props} />)

		component
			.find('button')
			.at(2)
			.simulate('click')

		expect(EditorUtil.renamePage).toHaveBeenCalledWith(6, '(Unnamed Module)')
	})

	test('EditorNav component clicks Rename Module button', () => {
		EditorUtil.getOrderedList.mockReturnValueOnce([
			{
				id: 6,
				type: 'heading',
				label: 'label6',
				flags: {
					assessment: true
				}
			}
		])
		jest.spyOn(window, 'prompt')
		window.prompt.mockReturnValueOnce('mockNewName')

		const props = { navState: {} }
		const component = mount(<EditorNav {...props} />)

		component
			.find('button')
			.at(2)
			.simulate('click')

		expect(EditorUtil.renamePage).toHaveBeenCalledWith(6, 'mockNewName')
	})

	test('EditorNav component clicks Copy URL button', () => {
		EditorUtil.getOrderedList.mockReturnValueOnce([])

		const props = { navState: {} }
		const component = mount(<EditorNav {...props} />)

		component
			.find('button')
			.at(3)
			.simulate('click')

		expect(ClipboardUtil.copyToClipboard).toHaveBeenCalled()
	})
})
