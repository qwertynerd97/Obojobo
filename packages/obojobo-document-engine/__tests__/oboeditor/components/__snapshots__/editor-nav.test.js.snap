// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorNav EditorNav component 1`] = `
<div
  className="viewer--components--nav editor--components--nav  is-not-locked is-not-open is-enabled"
>
  <ul />
  <div
    className="button-bar"
  >
    <button
      className="content-add-button"
      onClick={[Function]}
    >
      + Add Page
    </button>
    <button
      className="content-add-button"
      onClick={[Function]}
    >
      + Add Assessment
    </button>
    <br />
    <button
      className="content-add-button"
      onClick={[Function]}
    >
      Rename Module
    </button>
    <button
      className="content-add-button"
      onClick={[Function]}
    >
      Copy Module URL
    </button>
  </div>
</div>
`;

exports[`EditorNav EditorNav component with all list item types 1`] = `
<div
  className="viewer--components--nav editor--components--nav  is-locked is-not-open is-enabled"
>
  <ul>
    <li
      className="heading is-not-selected"
    >
      <span>
        label4
      </span>
    </li>
  </ul>
  <div
    className="button-bar"
  >
    <button
      className="content-add-button"
      onClick={[Function]}
    >
      + Add Page
    </button>
    <button
      className="content-add-button"
      onClick={[Function]}
    >
      + Add Assessment
    </button>
    <br />
    <button
      className="content-add-button"
      onClick={[Function]}
    >
      Rename Module
    </button>
    <button
      className="content-add-button"
      onClick={[Function]}
    >
      Copy Module URL
    </button>
  </div>
</div>
`;
